[Yul] Expression joiner.

chriseth opened the pull request:
Depends on https://github.com/ethereum/solidity/pull/5203

chriseth committed:
Expression joiner.

chriseth commented on libjulia/optimiser/ExpressionUnbreaker.cpp:
```diff
@@ -0,0 +1,166 @@
+	// Also we only descend into function calls if everything
+	// on the right is an identifier or literal.
+
+	size_t numVarsOrLitsOnRight = 0;
```
If someone has a suggestion about how to simplify this, please be my guest :)

leonardoalt commented on libjulia/optimiser/ExpressionUnbreaker.cpp:
```diff
@@ -0,0 +1,160 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimiser component that undoes what the ExpressionUnbreaker did, i.e.
```
What the `ExpressionBreaker` did

leonardoalt commented:
`Unbreaker` sounds a bit weird to me. I understand that the goal is to keep it connected to the `ExpressionBreaker` as the inverse operation, just wanted to bring that up.
Would you consider other names, such as `ExpressionInliner`?

leonardoalt commented on libjulia/optimiser/README.md:
```diff
@@ -56,6 +56,44 @@ depend on any side-effects.
+function calls can only appear in the right-hand side of a variable declaration,
+assignments or expression statements and all arguments have to be constants or variables.
+
+The benefits of this form is that it is much easier to re-order the sequence of opcodes
```
`benefits ... are`

leonardoalt commented on test/libjulia/ExpressionUnbreaker.cpp:
```diff
@@ -0,0 +1,326 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Unit tests for the expression breaker.
```
`unbreaker`

leonardoalt commented on test/libjulia/ExpressionUnbreaker.cpp:
```diff
@@ -0,0 +1,326 @@
+			let a := mload(2)
+			let b := mload(a)
+			a := 4
+		})"
```
Why is there no inlining in this case?

leonardoalt commented on test/libjulia/ExpressionUnbreaker.cpp:
```diff
@@ -0,0 +1,326 @@
+				a := mload(2)
+				x := add(a, 3)
+			}
+		})"
```
Why is there no inlining in this case?

leonardoalt commented on libjulia/optimiser/ExpressionUnbreaker.cpp:
```diff
@@ -0,0 +1,160 @@
+			assertThrow(varDecl.value, OptimizerException, "");
+
+			_e = std::move(*varDecl.value);
+			*latestStatement() = Block();
```
Why is this necessary?

chriseth commented:
Updated and renamed. I also made the constructor private and added a static function to prevent mis-use - not sure if it helps much, since the visitor functions have to be public.

chriseth commented:
Will rebase onto #5203
